execinfra/sql: return ContentionEvents through ProducerMetadata and add metric

asubiotto opened the pull request:
See individual commits for details. This PR adds a `sql.distsql.contended_queries.count` metric that is incremented if a query produces at least one contention event.

Release note: None

Closes #56798

RaduBerinde reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/57298#-:-MNUANO76AI5LcUiUMUk:bnfp4nl)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @RaduBerinde)



<!-- Sent from Reviewable.io -->

asubiotto commented:
Getting an issue because the new metric is not present in the chart catalog. Do I want it to be? I'm not sure what this means. I'll look into it tomorrow.

Commit:
execinfrapb: add roachpb.ContentionEvents to ProducerMetadata

This allows remote components to send back contention events through the
metadata protobuf.

Release note: None

Commit:
sql: return ContentionEvents in components and report through global metric

This commit changes KV layer-aware components to return any ContentionEvents
produced by the KV layer through metadata. Returning ContentionEvents in write
transactions is left for a future commit.

A global contention timeseries metric is also added and incremented at most
once per query by the distsql receiver if at least one contention event is
encountered.

Release note: None (change is not observable by the user)

asubiotto commented:
TFTR

bors r=RaduBerinde
